### PR TITLE
[@hapipal/toys] add initial types

### PR DESCRIPTION
**Pull request by timcosta**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## hapipal__toys ([<kbd>unpkg</kbd>](https://unpkg.com/browse/hapipal__toys@latest/))
was missing the following properties:
1. onPreStart
1. onPostStart
1. onPreStop
1. onPostStop
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1c80405694e2f457b268e7caae37a65745010efa
</p>

File: types/hapipal__toys/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+export function rootRealm(realm: ServerRealm): ServerRealm;
+
+export function state(realm: ServerRealm, pluginName: string): object;
+export function rootState(realm: ServerRealm, pluginName: string): object;
+
+export type AncestorRealmIterator = (realm: ServerRealm) => void;
+export function forEachAncestorRealm(realm: ServerRealm, fn: AncestorRealmIterator): void;
+
+export function asyncStorage(identifier: string): any;
```
Comment (Nargonath) [inline types/hapipal__toys/index.d.ts:99]: Perhaps it could be interesting to use a generic here too instead of `any` as the return type?

**Review by Nargonath — COMMENTED** [review]
Just a thing I noticed but I didn't look through all of it yet

File: types/hapipal__toys/index.d.ts (hunk 1 above)
Comment (timcosta) [reply types/hapipal__toys/index.d.ts:99]: I had originally written this as `export function asyncStorage<T>(identifier: string): T;` which worke better, but the linter complained about a single use generic.

**Comment by timcosta** [comment]
The Danger comment above is a little confusing - it's checking `hapipal__toys` instead of `@hapipal/toys` in the url which is causing a 404. Those items are definitely exported from the package though at https://unpkg.com/browse/@hapipal/toys@latest/lib/index.js.

File: types/hapipal__toys/index.d.ts (hunk 1 above)
Comment (timcosta) [reply types/hapipal__toys/index.d.ts:99]: https://github.com/Microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md at the bottom is the case mentioned. I could disable that rule if that seems better, as I agree that the `any` is a little weird in this use case.

**Review by amcasey — APPROVED** [review]
I think following the lint rule is preferable.

**Comment by timcosta** [comment]
Ready to merge

**Pull request merged**